# 4. 예외

JdbcTemplate을 대표로 하는 스프링의 데이터 액세스 기능에 담겨있는 예외처리와 관련된 접근 방법

# 4.1  사라진 SQLException

```java
public void deleteAll() throws SQLException {
	this.jdbcContext.executeSql("delete from users");
}

// JdbcTemplate 적용 전
// ---------------------------------------------------
// JdbcTemplate 적용 후 

pubic void deleteAll() {
	this.jdbcTemplate.update("delete from users");
}
```

# 4.1.1  초난감 예외 처리

## 예외 블랙홀

예외가 발생하면 그것을 catch 블록을 써서 잡아내는 것까지는 좋은데 그리고 아무것도 하지 않고 별문제 없는 것처럼 넘어가 버리는 건 정말 위험한 일이다.

```java
/* 1. 예외를 잡고 아무것도 하지 않는 경우 */
try {
...
} catch (SQLException e) {
}
/* 2. 출력만 하는 경우 */
} catch (SQLException e) {
	e.printStackTrace();
}
```

예외를 처리할 때 반드시 지켜야 할 핵심 원칙은 한가지다. 모든 예외는 적절하게 복구되든지 아니면 작업을 중단시키고 운영자 또는 개발자에게 분명하게 통보돼야 한다.

## 무의미하고 무책임한 throws

모든 메소드에 기계적으로 throws Exception을 붙이는 것은 예외 블랙홀보다는 조금 낫긴 하지만 이런 무책임한 throws 선언도 심각한 문제점이 있다.

그런 메소드 선언에서는 의미 있는 정보를 얻을 수 없다. 

⇒ 실행중에 예외적인 상황이 발생할 수 있다는 것인지, 그냥 습관적으로 복사해서 붙여 놓은 것인지 알 수 없다.

⇒ 어쩔 수 없이 메소드를 사용하는 메소드에도 throws Exception 을 붙일 수 밖에 없다.

결과적으로 적절한 처리를 통해 복구될 수 있는 예외상황도 제대로 다룰 수 있는 기회를 박탈 당한다. 

# 4.1.2 예외의 종류와 특징

자바에서 throw를 통해 발생시킬 수 있는 예외는 크게 세 가지가 있다.

- Error
    
    java.lang.Error 클래스의 서브클래스들로,시스템에 뭔가 비정상적인 상황이 발생했을 경우에 사용되므로 주로 자바 VM에서 발생시키는 것이고, 애플리케이션 코드에서 잡으려고 하면 안 된다.
    
    OutOfMemoryError, ThreadDeath 같은 에러는 catch블록으로 잡아도 아무런 대응 방법이 없다.
    
- Exception과 Checked Exception
    
    java.lang.Exception 클래스와 그 서브클래스들로 다시 Checked Exception 과 Unchecked Exception으로 구분된다.
    
    - Checked Exception
        
        Exception 클래스의 서브클래스이면서 RuntimeException 클래스를 상속하지 않는 것들
        
    - Unchecked Exception
        
        Exception 클래스의 서브클래스이면서 RuntimeException 클래스를 상속한 것들
        
    
    일반적으로 예외라고 하면 Checked Exception만을 말한다고 생각해도 되는데, Checked Exception이 발생할 수 있는 메소드를 사용할 경우 반드시 예외를 처리하는 코드를 함께 작성해야 한다. ⇒ 컴파일 에러 발생
    
- RuntimeException과 Unchecked Exception
    
    java.lang.RuntimeException 클래스를 상속한 예외들로, 피할 수 있지만 개발자가 부주의해서 발생할 수 있는 경우에 발생하도록 만든 것이 RuntimeException
    
    RuntimeException은 catch로 잡거나 throws로 선언하지 않아도 되지만 명시적으로 잡아도 상관없다.
    
    주로 프로그램의 오류가 있을 때 발생하도록 의도된 것들이다.
    
    NullPointerException, IllegalArgumentException 등
    

# 4.1.3 예외처리 방법

## 예외 복구

예외상황을 파악하고 문제를 해결해서 정상 상태로 돌려놓는 것

Checked Exception들은 예외를 어떤 식으로든 복구할 가능성이 있는 경우에 사용한다.

## 예외처리 회피

예외를 자신이 담당하지 않고 자신을 호출한 쪽으로 던져버리는 것

throw 문으로 선언해서 예외가 발생하면 알아서 던져지게 하거나 catch문으로 일단 예외를 잡은 후에 로그를 남기고 rethrow 하는 것

빈 catch 블록으로 예외를 잡는 것은 예외처리를 회피한 것이 아니다

예외를 회피하는 것은 예외를 복구하는 것처럼 의도가 분명해야 한다. 콜백/템플릿처럼 긴밀한 관계에 있는 다른 오브젝트에게 예외처리 책임을 분명히 지게 하거나, 자신을 사용하는 쪽에서 예외를 다루는 게 최선의 방법이라는 분명한 확신이 있어야 한다.

## 예외 전환 (Exception Translation)

예외를 복구해서 정상적인 상태로 만들 수 없기 때문에 예외를 메소드 밖으로 던지는데, 그대로 던지는 것이 아니라 적절한 예외로 전환해서 던진다

1. 내부에서 발생한 예외를 그대로 던지는 것이 그 예외상황에 대한 적절한 의미를 부여해주지 못하는 경우에, 의미를 분명하게 해줄 수 있는 예외로 바꿔서 던진다
    
    SQLException → DuplicateUserIdException
    
    보통 전환하는 예외에 원래 발생한 예외를 담에서 Nested Exception으로 만드는 것이 좋다
    
2. 예외를 처리하기 쉽고 단순하게 만들기 위해 포장(wrap)
    
    Nested Exception을 이용해 새로운 예외를 만들고 원인이 되는 예외를 내부에 담에서 던지는 방식은 같지만, 주로 예외 처리를  강제하는 Checked Exception을 RuntimeException으로 바꾸는 경우에 사용한다.
    
    Runtime Exception으로 변경되면 해당 예외가 발생하는 클라이언트에서 일일이 예외를 잡거나 다시 던지는 수고를 할 필요가 없다.
    

애플리케이션 로직 상에서 예외조건이 발견되거나 예외상황이 발생할 때 던지는 예외는 API가 던지는 예외가 아니라 애플리케이션 코드에서 의도적으로 던지는 예외이므로, 이때는 체크 예외를 사용하는 것이 적절하다.

일반적으로 Checked Exception을 계속 throws를 사용해 넘기는 것은 무의미하다.

어차피 복구가 불가능한 예외라면 가능한 한 빨리 런타임 예외로 포장해 던지게 해서 다른 계층의 메소드를 작성할 때 불필요한 throws 선언이 들어가지 않도록 해줘야 한다.